source/extensions/filters/common/expr/evaluator.cc

```diff
@@ -138,14 +153,127 @@ BuilderInstanceSharedPtr createBuilder(Protobuf::Arena* arena) {
     throw CelException(absl::StrCat("failed to register extension regex functions: ",
                                     ext_register_status.message()));
   }
-  return std::make_shared<BuilderInstance>(std::move(builder));
+  // Register string extension functions only if enabled in configuration.
+  if (enable_string_functions) {
+    auto string_register_status =
+        cel::extensions::RegisterStringsFunctions(builder->GetRegistry(), options);
+    if (!string_register_status.ok()) {
+      throw CelException(absl::StrCat("failed to register extension string functions: ",
+                                      string_register_status.message()));
+    }
+  }
+  return builder;
 }
 
-SINGLETON_MANAGER_REGISTRATION(expression_builder);
+// Nested map to store builders for different configurations.
+// The outer map is a singleton, and the inner map stores builders by configuration hash.
+class BuilderCache : public Singleton::Instance {
+public:
+  using ConfigHash = size_t;
+
+  BuilderInstanceSharedConstPtr
+  getOrCreateBuilder(const envoy::config::core::v3::CelExpressionConfig* config) {
+    ConfigHash hash = 0;
+    if (config) {
+      // Create a simple hash based on the configuration values.
+      hash = absl::HashOf(config->enable_string_conversion(), config->enable_string_concat(),
+                          config->enable_string_functions(), config->enable_constant_folding());
+    }
+
+    absl::MutexLock lock(&mutex_);
+    auto it = builders_.find(hash);
+    if (it != builders_.end()) {
+      return it->second;
+    }
+
+    // Create new builder with the configuration.
+    auto builder = createBuilder(nullptr, config);
+    auto instance = std::make_shared<BuilderInstance>(std::move(builder));
+    builders_[hash] = instance;
+    return instance;
+  }
+
+private:
+  absl::Mutex mutex_;
+  absl::flat_hash_map<ConfigHash, BuilderInstanceSharedConstPtr> builders_ ABSL_GUARDED_BY(mutex_);
```

Comment:
This is a common pattern - it should be `weak_ptr` here, and put a TODO to clean up null weak_ptr at some point.